### PR TITLE
[VPlan] Support live-ins without underlying IR in type analysis.

### PR DESCRIPTION
**Pull request by fhahn**
A VPlan contains multiple live-ins without underlying IR, like VFxUF or VectorTripCount. Trying to infer the scalar type of those causes a crash at the moment.

Update VPTypeAnalysis to take a VPlan in its constructor and assign types to those live-ins up front. All those live-ins share the type of the canonical IV.

### CODE REVIEW

**Comment by fhahn** [comment]
ping :)

File: llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp [hunk 1]
```diff
@@ -203,6 +198,16 @@ Type *VPTypeAnalysis::inferScalarTypeForRecipe(const VPReplicateRecipe *R) {
...
+  CachedTypes[&Plan.getVectorTripCount()] = CanIVTy;
+  CachedTypes[&Plan.getVFxUF()] = CanIVTy;
+  CachedTypes[Plan.getTripCount()] = CanIVTy;
+  if (auto *BTC = Plan.getBackedgeTakenCount())
+    CachedTypes[BTC] = CanIVTy;
+}
+
 Type *VPTypeAnalysis::inferScalarType(const VPValue *V) {
   if (Type *CachedTy = CachedTypes.lookup(V))
     return CachedTy;
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:203]: Updated as suggested, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp [hunk 2]
```diff
@@ -35,12 +35,7 @@ Type *VPTypeAnalysis::inferScalarTypeForRecipe(const VPInstruction *R) {
     CachedTypes[OtherV] = ResTy;
     return ResTy;
   }
-  case Instruction::ICmp: {
-    // TODO: Check if types for both operands agree. This also requires
-    // type-inference for the vector-trip-count, which is missing at the moment.
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:40]: Yep.

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 3]
```diff
@@ -2929,6 +2927,9 @@ class VPlan {
     return BackedgeTakenCount;
   }
 
+  /// Return the backedge taken count of the original loop, if set.
+  VPValue *getBackedgeTakenCount() { return BackedgeTakenCount; }
+
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2932]:
```suggestion
```
is this still needed?

File: llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp [hunk 4]
```diff
@@ -207,8 +202,11 @@ Type *VPTypeAnalysis::inferScalarType(const VPValue *V) {
   if (Type *CachedTy = CachedTypes.lookup(V))
     return CachedTy;
 
-  if (V->isLiveIn())
-    return V->getLiveInIRValue()->getType();
+  if (V->isLiveIn()) {
+    if (auto *IRValue = V->getLiveInIRValue())
+      return IRValue->getType();
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:207]: nit: worth a comment to explain who's expected to use CanonicalIVTy.

File: llvm/lib/Transforms/Vectorize/VPlanAnalysis.h [hunk 5]
```diff
@@ -15,6 +15,7 @@ namespace llvm {
 
 class LLVMContext;
 class VPValue;
+class VPlan;
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanAnalysis.h:18]:
```suggestion
```
is this still needed?

File: llvm/lib/Transforms/Vectorize/VPlanAnalysis.h [hunk 6]
```diff
@@ -35,6 +36,7 @@ class Type;
 /// of the previously inferred types.
 class VPTypeAnalysis {
   DenseMap<const VPValue *, Type *> CachedTypes;
+  Type *CanonicalIVTy;
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanAnalysis.h:39]: nit: worth a comment what this is for.

File: llvm/lib/Transforms/Vectorize/VPlanAnalysis.h [hunk 7]
```diff
@@ -47,7 +49,8 @@ class VPTypeAnalysis {
   Type *inferScalarTypeForRecipe(const VPReplicateRecipe *R);
 
 public:
-  VPTypeAnalysis(LLVMContext &Ctx) : Ctx(Ctx) {}
+  VPTypeAnalysis(Type *CanonicalIVTy, LLVMContext &Ctx)
+      : CanonicalIVTy(CanonicalIVTy), Ctx(Ctx) {}
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanAnalysis.h:53]: nit (unrelated to this patch): does VPTypeAnalysis really need `Ctx`?

**Review by ayalz — APPROVED** [review]
Thanks for accommodating, this looks good to me, adding a couple of nits, worth explaining what CanonicalIVTy is used for, possibly along with a TODO to promote the temporary live-ins-w/o-underlying-value (which need it) into recipes.

File: llvm/lib/Transforms/Vectorize/VPlanAnalysis.h (hunk 7 above)
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/VPlanAnalysis.h:55]: Is there any current use of `Ctx`? Sure, if this is indeed dead, should be eliminated separately.

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanAnalysis.h:55]: The current uses are to create `i1` & `void` types:

https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp#L63
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp#L163
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp#L189

**Pull request merged**